### PR TITLE
New functions

### PR DESCRIPTION
**Pull request by DeLaGuardo**
* `walk` for traversing into data structure and apply supplied function to leaf
  values
* `composition` - same as `clojure.core/comp` but for axel-f functions
* `validate.presence` to check some value to exists and throw otherwise
* `validate.not-empty` to check if the value is not empty and throw otherwise
* `coerce.to-%` set of coercers to change value type

### CODE REVIEW

File: src/axel_f/excel/coerce.cljc [hunk 1]
```diff
@@ -27,3 +28,62 @@
     true "TRUE"
     false "FALSE"
     (str item)))
+
+(defn to-string*
+  "Try to coerce given value to a string type. Returns null for empty value."
```
Comment (DG4227) [inline src/axel_f/excel/coerce.cljc]: Most other axel-f functions use present tense verbs - 'rounds', 'returns', 'subtracts' etc. The first half of this string would be a better fit as 'Tries to coerce given value to a string type.'

File: src/axel_f/excel/coerce.cljc [hunk 2]
```diff
@@ -27,3 +28,62 @@
...
+  [^{:doc "Any object to coerce to a string."} obj]
+  (when obj (str obj)))
+
+(def to-string #'to-string*)
+
+(defn to-integer*
+  "Try to coerce given value to an integer type. Returns null for empty or not reducible to an integer type value."
```
Comment (DG4227) [inline src/axel_f/excel/coerce.cljc]: `Returns null for empty or not reducible to an integer type value.` is a bit confusing. Is 'not reducible' the error response for an integer type? Or is there a general error 'Not reducible to an integer type'? 

'Tries to coerce given value to an integer type. '

File: src/axel_f/excel/coerce.cljc [hunk 3]
```diff
@@ -27,3 +28,62 @@
...
+                   nil
+                   n))))
+    (catch #?(:clj java.lang.NumberFormatException
+              :cljs js/Error) _
+      nil)))
+
+(def to-integer #'to-integer*)
+
+(defn to-float*
+  "Try to coerce given value to float type. Returns null for empty or not reducible to a float type value."
```
Comment (DG4227) [inline src/axel_f/excel/coerce.cljc]: See above^^

File: src/axel_f/excel/coerce.cljc [hunk 4]
```diff
@@ -27,3 +28,62 @@
...
+                   nil
+                   n))))
+    (catch #?(:clj java.lang.NumberFormatException
+              :cljs js/Error) _
+      nil)))
+
+(def to-float #'to-float*)
+
+(defn to-boolean*
+  "Try to coerce given value to a boolean type. Returns null for empty or not reducible to a boolean type value."
```
Comment (DG4227) [inline src/axel_f/excel/coerce.cljc]: ^^

File: src/axel_f/excel/validate.cljc [hunk 5]
```diff
@@ -0,0 +1,28 @@
+(ns axel-f.excel.validate)
+
+(defn presence*
+  "Validate and return if not null given value and throw an Exception with given message (default is 'Argument is required')."
```
Comment (DG4227) [inline src/axel_f/excel/validate.cljc]: This one is a bit hard to understand - does it mean 'Validate and return given value if not null, and throw an exception with given message'?

File: src/axel_f/excel/validate.cljc [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  [^{:doc "Object to check for presence."} x
+   & [^{:doc "Optional message to be thrown."} msg]]
+  (let [msg (or msg "Argument required")]
+    (if (some? x) x (throw (ex-info msg {:type ::validate-error
+                                         :subtype ::presence})))))
+
+(def presence #'presence*)
+
+(defn not-empty**
+  "Validate and return if not empty given value and throw an Exception with given message (default is 'Argument can not be empty'). "
```
Comment (DG4227) [inline src/axel_f/excel/validate.cljc]: same as above ^^

File: src/axel_f/excel/coerce.cljc (hunk 2 above)
Comment (DeLaGuardo) [reply src/axel_f/excel/coerce.cljc]: I wanted to outline that null will be a result of `to-integer('foo')`. Underling function will throw an exception but the function catches and transforms it to null.

File: src/axel_f/excel/validate.cljc (hunk 5 above)
Comment (DeLaGuardo) [reply src/axel_f/excel/validate.cljc]: exactly as you phrase it, thanks!

File: src/axel_f/excel/coerce.cljc [hunk 7]
```diff
@@ -27,3 +28,62 @@
...
+
+(defn to-string*
+  "Tries to coerce given value to a string type. Returns null for empty value."
+  [^{:doc "Any object to coerce to a string."} obj]
+  (when obj (str obj)))
+
+(def to-string #'to-string*)
+
+(defn to-integer*
+  "Tries to coerce given value to an integer type. Returns null for empty or not reducible to an integer type value."
```
Comment (DG4227) [inline src/axel_f/excel/coerce.cljc]: For this and subsequent cases, I'd try to rephrase the second half. 

Maybe something like:

> Returns null for empty values or values not reducible to integer

Comment (DeLaGuardo) [reply src/axel_f/excel/coerce.cljc]: thanks!

**Pull request merged**